bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,414 @@
+            self._workers: weakref.WeakSet = weakref.WeakSet()
+            for _ in range(self._min_workers):  # fire up stable worker pool
+                # careful with back references to self!
+                worker = threading.Thread(
+                    target=AutoScalingThreadPool._do_work, args=(weak_self, self._logger, False), daemon=True
+                )
+                worker.start()
+                self._workers.add(worker)
+            # register workers for external joining on interpreter shutdown
+            AutoScalingThreadPool._all_workers.update(self._workers)
+
+        self._waitqueues: typing.Dict[typing.Callable[..., typing.Any], collections.deque] = collections.defaultdict(
+            collections.deque
+        )
+        self._runlists: typing.Dict[typing.Callable[..., typing.Any], typing.List[AutoScalingThreadPool.Work]] = (
+            collections.defaultdict(list)
+        )
+        self._runslots = threading.Semaphore(0)
+
+    @property
+    def workers(self) -> typing.List[threading.Thread]:
+        with self._submit_lock:
+            # avoid races and proactively drop shutdown workers
+            self._workers = weakref.WeakSet([w for w in self._workers if w.is_alive()])
+            return list(self._workers)
+
+    @property
+    def working(self) -> bool:
+        with self._submit_lock:
+            return len(self._runlists) > 0 or len(self._waitqueues) > 0
+
+    @property
+    def capped(self) -> bool:
+        with self._submit_lock:
+            return len(self._waitqueues) > 0
+
+    def wait(self, timeout: typing.Optional[float] = None) -> bool:
+        with self._submit_lock:
+            futures = [work.future for runlist in self._runlists.values() for work in runlist]
+            futures += [work.future for waitqueue in self._waitqueues.values() for work in waitqueue]
+        done, not_done = concurrent.futures.wait(futures, timeout=timeout)
+        return len(not_done) == 0
+
+    def _cleanup_after(self, work: "AutoScalingThreadPool.Work") -> bool:
+        complete = True
+        with self._submit_lock:
+            self._runlists[work.fn].remove(work)
+            if work.fn in self._waitqueues and self._waitqueues[work.fn]:  # continue with pending work
+                work = self._waitqueues[work.fn].popleft()
+                while work.cancelled() and self._waitqueues[work.fn]:
+                    work.notify_cancelation()
+                    work = self._waitqueues[work.fn].popleft()
+                if not self._waitqueues[work.fn]:
+                    del self._waitqueues[work.fn]
+                if not work.cancelled():
+                    self._runlists[work.fn].append(work)
+                    self._runqueue.put(work)  # actually submit work
+                    complete = False
+                else:
+                    work.notify_cancelation()
+            if not self._runlists[work.fn]:
```

Comment:
Same for this one. I'm not sure if it's a hard guarantee or not, so feel free to correct me.

Reply:
In this case it is an empty `list`.